projects/ngx-pattern/src/lib/ngx-pattern.directive.ts

```diff
@@ -1,24 +1,44 @@
 import {
   Directive,
   ElementRef,
-  HostListener, Inject,
+  HostListener,
+  Inject,
   Input,
   OnChanges,
-  SimpleChanges,
+  OnDestroy,
+  OnInit,
+  Optional
 } from '@angular/core';
+import { NgControl } from '@angular/forms';
 
 @Directive({
   selector: '[ngxPattern]'
 })
-export class NgxPatternDirective implements OnChanges {
+export class NgxPatternDirective implements OnChanges, OnInit, OnDestroy {
   @Input() ngxPattern: RegExp | string;
 
   private regex: RegExp;
+  private lastSelectionStart = 0;
+  private lastSelectionEnd = 0;
+  private lastValue = '';
+  onPasteHandler: (e: ClipboardEvent) => void;
+  onKeydownHandler: (e: KeyboardEvent) => void;
 
-  constructor(@Inject(ElementRef) private host: ElementRef) {
+  constructor(@Inject(ElementRef) private host: ElementRef, @Optional() @Inject(NgControl) private control: NgControl) {
   }
 
-  ngOnChanges(changes: SimpleChanges): void {
+  ngOnInit(): void {
+    this.onPasteHandler = (e: ClipboardEvent) => {
+      this.onPaste(e);
+    };
+    this.onKeydownHandler = (e: KeyboardEvent) => {
+      this.onKeyDown(e);
+    };
+    this.host.nativeElement.addEventListener('keydown', this.onKeydownHandler);
```

Comment:
Can you implement this using `fromEvent` so it's angular friendly? Then you can unsubscribe on destroy using `takeUntil` and a subject or just remember the subscriptions in an array and call unsubscribe. _Also, this way you won't need the callbacks you had to write above._

Reply:
Sure. Thanks for the comment. I've done the corrections. Check.